worker/default_lifecycle.go

```
@@ -0,0 +1,227 @@
+	errs chan error
+
+	// closer is a channel that receives an empty message when it is time to
+	// close.
+	closer chan struct{}
+
+	// wg is a WaitGroup that keeps track of all actively owned tasks. When
+	// a task is pulled, the state of this WaitGroup increases, and
+	// conversely decreases when the task is either COMPLETED, or FAILED.
+	wg sync.WaitGroup
+}
+
+// NewLifecycle allocates and returns a pointer to a new instance of
+// DefaultLifecycle. It uses the specified pool to make connections into Redis
+// and takes a source, which is the keyspace of tasks that need work done, and
+// id is the ID of this particular worker.
+func NewLifecycle(pool *redis.Pool, source, id string) *DefaultLifecycle {
+	workerName := fmt.Sprintf("%s:worker_%s", source, id)
+
+	return &DefaultLifecycle{
+		pool: pool,
+
+		availableTasks: queue.NewByteQueue(pool, source),
+		workingTasks:   queue.NewDurableQueue(pool, source, workerName),
+	}
+}
+
+var _ Lifecycle = new(DefaultLifecycle)
+
+func (l *DefaultLifecycle) Await() {
+	l.wg.Wait()
+}
+
+// Listen returns a channel of tasks and error that are pulled from the
+// main processing queue. Once a *Task is able to be read from the <-chan *Task,
+// that Task is ready to be worked on and is in the appropriate locations in
+// Redis. StopListening() can be called to terminate.
+func (l *DefaultLifecycle) Listen() (<-chan *Task, <-chan error) {
+	l.tasks = make(chan *Task)
+	l.errs = make(chan error)
+	l.closer = make(chan struct{})
+
+	go l.recv()
+
+	return l.tasks, l.errs
+}
+
+// StopListening closes the queue "pull" task at the next opportunity. Tasks
+// can still be marked as completed or abandoned, but no new tasks will
+// be generated.
+func (l *DefaultLifecycle) StopListening() {
+	l.closer <- struct{}{}
+}
+
+// Complete marks a task as having been completed, removing it from the
+// worker's queue.
+func (l *DefaultLifecycle) Complete(task *Task) error {
+	l.removeTask(task)
+
+	return nil
```

Probably want to return the error from removeTask